crates/database/src/settlement_observations.rs

```diff
@@ -0,0 +1,84 @@
+use {crate::events::EventIndex, bigdecimal::BigDecimal, sqlx::PgConnection};
+
+#[derive(Debug, PartialEq, sqlx::FromRow)]
+pub struct Observation {
+    pub gas_used: BigDecimal,
+    pub effective_gas_price: BigDecimal,
+    pub surplus: BigDecimal,
+    pub fee: BigDecimal,
+    pub block_number: i64,
+    pub log_index: i64,
+}
+
+pub async fn upsert(
+    ex: &mut PgConnection,
+    event: &EventIndex,
+    gas_used: BigDecimal,
+    effective_gas_price: BigDecimal,
+    surplus: BigDecimal,
+    fee: BigDecimal,
+) -> Result<(), sqlx::Error> {
+    const QUERY: &str = r#"
+    INSERT INTO settlement_observations (gas_used, effective_gas_price, surplus, fee, block_number, log_index) 
+    VALUES ($1, $2, $3, $4, $5, $6)
+    ON CONFLICT (block_number, log_index) DO UPDATE SET gas_used = $1, effective_gas_price = $2, surplus = $3, fee = $4;
+    "#;
+    sqlx::query(QUERY)
+        .bind(gas_used)
+        .bind(effective_gas_price)
+        .bind(surplus)
+        .bind(fee)
+        .bind(event.block_number)
+        .bind(event.log_index)
+        .execute(ex)
+        .await?;
+    Ok(())
+}
+
+#[cfg(test)]
+mod tests {
+    use {super::*, sqlx::Connection};
+
+    // helper function to make roundtrip possible
+    pub async fn fetch(
+        ex: &mut PgConnection,
+        event: &EventIndex,
+    ) -> Result<Option<Observation>, sqlx::Error> {
+        const QUERY: &str =
+            r#"SELECT * FROM settlement_observations WHERE block_number = $1 AND log_index = $2"#;
+        sqlx::query_as(QUERY)
+            .bind(event.block_number)
+            .bind(event.log_index)
+            .fetch_optional(ex)
+            .await
+    }
+
+    #[tokio::test]
+    #[ignore]
+    async fn postgres_roundtrip() {
```

Comment:
I think there should also be a test demonstrating the "ON CONFLICT" aspect of the insertion query.